### ISSUE TITLE
Cannot dump in skore project scikit-learn models

### ISSUE THREAD

**Issue by glemaitre** [probabl-ai/skore#954]
When executing the example:

https://skore.probabl.ai/0.5/auto_examples/00_getting_started/plot_01_getting_started.html#sphx-glr-auto-examples-00-getting-started-plot-01-getting-started-py

Putting the grid-search in the skore project will trigger a `NotImplementedError`.

```pytb
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[8], line 1
----> 1 my_project.put("my_gs_cv", gs_cv)
      2 my_project.put("my_df", df)
      3 my_project.put("my_fig", fig)

File ~/miniforge3/envs/xxx/lib/python3.13/site-packages/skore/project/project.py:142, in Project.put(self, key, value)
    139 if not isinstance(key, str):
    140     raise TypeError(f"Key must be a string (found '{type(key)}')")
--> 142 self.item_repository.put_item(key, object_to_item(value))

File ~/miniforge3/envs/xxx/lib/python3.13/site-packages/skore/item/__init__.py:46, in object_to_item(object)
     36     with suppress(ImportError, ItemTypeError):
     37         # ImportError:
     38         #     The factories are responsible to import third-party libraries in a
   (...)
     42         #     The factories are responsible for checking that parameters are of the
     43         #     correct type. If not, they throw a ItemTypeError exception.
     44         return cls.factory(object)
---> 46 raise NotImplementedError(f"Type '{object.__class__}' is not supported.")

NotImplementedError: Type '<class 'sklearn.model_selection._search.GridSearchCV'>' is not supported.
```

### Steps to reproduce

```shell
mamba create --name xxx python scikit-learn skore jupyter notebook pandas polars matplotlib
```

### List of package installed

```
# packages in environment at /Users/glemaitre/miniforge3/envs/xxx:
#
# Name                    Version                   Build  Channel
annotated-types           0.7.0              pyhd8ed1ab_1    conda-forge
anyio                     4.7.0              pyhd8ed1ab_0    conda-forge
appnope                   0.1.4              pyhd8ed1ab_1    conda-forge
argon2-cffi               23.1.0             pyhd8ed1ab_1    conda-forge
argon2-cffi-bindings      21.2.0          py313h20a7fcf_5    conda-forge
arrow                     1.3.0              pyhd8ed1ab_1    conda-forge
asttokens                 3.0.0              pyhd8ed1ab_1    conda-forge
async-lru                 2.0.4              pyhd8ed1ab_1    conda-forge
attrs                     24.2.0             pyh71513ae_1    conda-forge
aws-c-auth                0.8.0               h8bc59a9_15    conda-forge
aws-c-cal                 0.8.1                hc8a0bd2_3    conda-forge
aws-c-common              0.10.6               h5505292_0    conda-forge
aws-c-compression         0.3.0                hc8a0bd2_5    conda-forge
aws-c-event-stream        0.5.0               h54f970a_11    conda-forge
aws-c-http                0.9.2                h96aa502_4    conda-forge
aws-c-io                  0.15.3               haba67d1_4    conda-forge
aws-c-mqtt                0.11.0              h24f418c_12    conda-forge
aws-c-s3                  0.7.7                h1be5864_0    conda-forge
aws-c-sdkutils            0.2.1                hc8a0bd2_4    conda-forge
aws-checksums             0.2.2                hc8a0bd2_4    conda-forge
aws-crt-cpp               0.29.7               h19a973c_7    conda-forge
aws-sdk-cpp               1.11.458             he0ff2e4_4    conda-forge
azure-core-cpp            1.14.0               hd50102c_0    conda-forge
azure-identity-cpp        1.10.0               hc602bab_0    conda-forge
azure-storage-blobs-cpp   12.13.0              h7585a09_1    conda-forge
azure-storage-common-cpp  12.8.0               h9ca1f76_1    conda-forge
azure-storage-files-datalake-cpp 12.12.0              hcdd55da_1    conda-forge
babel                     2.16.0             pyhd8ed1ab_1    conda-forge
beautifulsoup4            4.12.3             pyha770c72_1    conda-forge
bleach                    6.2.0              pyhd8ed1ab_1    conda-forge
brotli                    1.1.0                hd74edd7_2    conda-forge
brotli-bin                1.1.0                hd74edd7_2    conda-forge
brotli-python             1.1.0           py313h3579c5c_2    conda-forge
bzip2                     1.0.8                h99b78c6_7    conda-forge
c-ares                    1.34.4               h5505292_0    conda-forge
ca-certificates           2024.12.14           hf0a4a13_0    conda-forge
cached-property           1.5.2                hd8ed1ab_1    conda-forge
cached_property           1.5.2              pyha770c72_1    conda-forge
certifi                   2024.8.30          pyhd8ed1ab_0    conda-forge
cffi                      1.17.1          py313hc845a76_0    conda-forge
charset-normalizer        3.4.0              pyhd8ed1ab_1    conda-forge
click                     8.1.7           unix_pyh707e725_1    conda-forge
colorama                  0.4.6              pyhd8ed1ab_1    conda-forge
comm                      0.2.2              pyhd8ed1ab_1    conda-forge
contourpy                 1.3.1           py313h0ebd0e5_0    conda-forge
cycler                    0.12.1             pyhd8ed1ab_1    conda-forge
debugpy                   1.8.11          py313h928ef07_0    conda-forge
decorator                 5.1.1              pyhd8ed1ab_1    conda-forge
defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge
diskcache                 5.6.3              pyhd8ed1ab_1    conda-forge
dnspython                 2.7.0              pyhff2d567_1    conda-forge
email-validator           2.2.0              pyhd8ed1ab_1    conda-forge
email_validator           2.2.0                hd8ed1ab_1    conda-forge
entrypoints               0.4                pyhd8ed1ab_1    conda-forge
exceptiongroup            1.2.2              pyhd8ed1ab_1    conda-forge
executing                 2.1.0              pyhd8ed1ab_1    conda-forge
fastapi                   0.115.6            pyhd8ed1ab_0    conda-forge
fastapi-cli               0.0.6              pyhd8ed1ab_0    conda-forge
filelock                  3.16.1             pyhd8ed1ab_1    conda-forge
fonttools                 4.55.3          py313ha9b7d5b_0    conda-forge
fqdn                      1.5.1              pyhd8ed1ab_1    conda-forge
freetype                  2.12.1               hadb7bae_2    conda-forge
fsspec                    2024.10.0          pyhd8ed1ab_1    conda-forge
gflags                    2.2.2             hf9b8971_1005    conda-forge
glog                      0.7.1                heb240a5_0    conda-forge
h11                       0.14.0             pyhd8ed1ab_1    conda-forge
h2                        4.1.0              pyhd8ed1ab_1    conda-forge
hpack                     4.0.0              pyhd8ed1ab_1    conda-forge
httpcore                  1.0.7              pyh29332c3_1    conda-forge
httptools                 0.6.4           py313h90d716c_0    conda-forge
httpx                     0.28.1             pyhd8ed1ab_0    conda-forge
huggingface_hub           0.26.5             pyhd8ed1ab_1    conda-forge
hyperframe                6.0.1              pyhd8ed1ab_1    conda-forge
icu                       75.1                 hfee45f7_0    conda-forge
idna                      3.10               pyhd8ed1ab_1    conda-forge
importlib-metadata        8.5.0              pyha770c72_1    conda-forge
importlib_resources       6.4.5              pyhd8ed1ab_1    conda-forge
ipykernel                 6.29.5             pyh57ce528_0    conda-forge
ipython                   8.30.0             pyh707e725_0    conda-forge
ipywidgets                8.1.5              pyhd8ed1ab_1    conda-forge
isoduration               20.11.0            pyhd8ed1ab_1    conda-forge
jedi                      0.19.2             pyhd8ed1ab_1    conda-forge
jinja2                    3.1.4              pyhd8ed1ab_1    conda-forge
joblib                    1.4.2              pyhd8ed1ab_1    conda-forge
json5                     0.10.0             pyhd8ed1ab_1    conda-forge
jsonpointer               3.0.0           py313h8f79df9_1    conda-forge
jsonschema                4.23.0             pyhd8ed1ab_1    conda-forge
jsonschema-specifications 2024.10.1          pyhd8ed1ab_1    conda-forge
jsonschema-with-format-nongpl 4.23.0               hd8ed1ab_1    conda-forge
jupyter                   1.1.1              pyhd8ed1ab_1    conda-forge
jupyter-lsp               2.2.5              pyhd8ed1ab_1    conda-forge
jupyter_client            8.6.3              pyhd8ed1ab_1    conda-forge
jupyter_console           6.6.3              pyhd8ed1ab_1    conda-forge
jupyter_core              5.7.2              pyh31011fe_1    conda-forge
jupyter_events            0.10.0             pyhd8ed1ab_1    conda-forge
jupyter_server            2.14.2             pyhd8ed1ab_1    conda-forge
jupyter_server_terminals  0.5.3              pyhd8ed1ab_1    conda-forge
jupyterlab                4.3.3              pyhd8ed1ab_0    conda-forge
jupyterlab_pygments       0.3.0              pyhd8ed1ab_2    conda-forge
jupyterlab_server         2.27.3             pyhd8ed1ab_1    conda-forge
jupyterlab_widgets        3.0.13             pyhd8ed1ab_1    conda-forge
kiwisolver                1.4.7           py313hf9c7212_0    conda-forge
krb5                      1.21.3               h237132a_0    conda-forge
lcms2                     2.16                 ha0e7c42_0    conda-forge
lerc                      4.0.0                h9a09cb3_0    conda-forge
libabseil                 20240722.0      cxx17_hf9b8971_1    conda-forge
libarrow                  18.1.0           h4a2f8bd_6_cpu    conda-forge
libarrow-acero            18.1.0           hf07054f_6_cpu    conda-forge
libarrow-dataset          18.1.0           hf07054f_6_cpu    conda-forge
libarrow-substrait        18.1.0           h86344ea_6_cpu    conda-forge
libblas                   3.9.0           25_osxarm64_openblas    conda-forge
libbrotlicommon           1.1.0                hd74edd7_2    conda-forge
libbrotlidec              1.1.0                hd74edd7_2    conda-forge
libbrotlienc              1.1.0                hd74edd7_2    conda-forge
libcblas                  3.9.0           25_osxarm64_openblas    conda-forge
libcrc32c                 1.1.2                hbdafb3b_0    conda-forge
libcurl                   8.11.1               h73640d1_0    conda-forge
libcxx                    19.1.5               ha82da77_0    conda-forge
libdeflate                1.22                 hd74edd7_0    conda-forge
libedit                   3.1.20191231         hc8eb9b7_2    conda-forge
libev                     4.33                 h93a5062_2    conda-forge
libevent                  2.1.12               h2757513_1    conda-forge
libexpat                  2.6.4                h286801f_0    conda-forge
libffi                    3.4.2                h3422bc3_5    conda-forge
libgfortran               5.0.0           13_2_0_hd922786_3    conda-forge
libgfortran5              13.2.0               hf226fd6_3    conda-forge
libgoogle-cloud           2.32.0               h8d8be31_0    conda-forge
libgoogle-cloud-storage   2.32.0               h7081f7f_0    conda-forge
libgrpc                   1.67.1               hc70892a_0    conda-forge
libiconv                  1.17                 h0d3ecfb_2    conda-forge
libjpeg-turbo             3.0.0                hb547adb_1    conda-forge
liblapack                 3.9.0           25_osxarm64_openblas    conda-forge
liblzma                   5.6.3                h39f12f2_1    conda-forge
libmpdec                  4.0.0                h99b78c6_0    conda-forge
libnghttp2                1.64.0               h6d7220d_0    conda-forge
libopenblas               0.3.28          openmp_hf332438_1    conda-forge
libparquet                18.1.0           h636d7b7_6_cpu    conda-forge
libpng                    1.6.44               hc14010f_0    conda-forge
libprotobuf               5.28.2               h8f0b736_0    conda-forge
libre2-11                 2024.07.02           h2348fd5_1    conda-forge
libsodium                 1.0.20               h99b78c6_0    conda-forge
libsqlite                 3.47.2               h3f77e49_0    conda-forge
libssh2                   1.11.1               h9cc3647_0    conda-forge
libthrift                 0.21.0               h64651cc_0    conda-forge
libtiff                   4.7.0                ha962b0a_2    conda-forge
libutf8proc               2.9.0                h5505292_1    conda-forge
libuv                     1.49.2               h7ab814d_0    conda-forge
libwebp-base              1.4.0                h93a5062_0    conda-forge
libxcb                    1.17.0               hdb1d25a_0    conda-forge
libxml2                   2.13.5               h178c5d8_1    conda-forge
libzlib                   1.3.1                h8359307_2    conda-forge
llvm-openmp               19.1.5               hdb05f8b_0    conda-forge
lz4-c                     1.10.0               h286801f_1    conda-forge
markdown-it-py            3.0.0              pyhd8ed1ab_1    conda-forge
markupsafe                3.0.2           py313ha9b7d5b_1    conda-forge
matplotlib                3.9.4           py313h39782a4_0    conda-forge
matplotlib-base           3.9.4           py313haaf02c0_0    conda-forge
matplotlib-inline         0.1.7              pyhd8ed1ab_1    conda-forge
mdurl                     0.1.2              pyhd8ed1ab_1    conda-forge
mistune                   3.0.2              pyhd8ed1ab_1    conda-forge
munkres                   1.1.4              pyh9f0ad1d_0    conda-forge
nbclient                  0.10.1             pyhd8ed1ab_0    conda-forge
nbconvert-core            7.16.4             pyhff2d567_2    conda-forge
nbformat                  5.10.4             pyhd8ed1ab_1    conda-forge
ncurses                   6.5                  h7bae524_1    conda-forge
nest-asyncio              1.6.0              pyhd8ed1ab_1    conda-forge
notebook                  7.3.1              pyhd8ed1ab_0    conda-forge
notebook-shim             0.2.4              pyhd8ed1ab_1    conda-forge
numpy                     2.2.0           py313ha4a2180_0    conda-forge
openjpeg                  2.5.3                h8a3d83b_0    conda-forge
openssl                   3.4.0                h39f12f2_0    conda-forge
orc                       2.0.3                hbcee414_1    conda-forge
overrides                 7.7.0              pyhd8ed1ab_0    conda-forge
packaging                 24.2               pyhd8ed1ab_2    conda-forge
pandas                    2.2.3           py313h47b39a6_1    conda-forge
pandocfilters             1.5.0              pyhd8ed1ab_0    conda-forge
parso                     0.8.4              pyhd8ed1ab_1    conda-forge
pexpect                   4.9.0              pyhd8ed1ab_1    conda-forge
pickleshare               0.7.5           pyhd8ed1ab_1004    conda-forge
pillow                    11.0.0          py313h97432e1_0    conda-forge
pip                       24.3.1             pyh145f28c_0    conda-forge
pkgutil-resolve-name      1.3.10             pyhd8ed1ab_2    conda-forge
platformdirs              4.3.6              pyhd8ed1ab_1    conda-forge
plotly                    5.24.1             pyhd8ed1ab_1    conda-forge
polars                    1.17.1          py313h8ea26c2_0    conda-forge
prometheus_client         0.21.1             pyhd8ed1ab_0    conda-forge
prompt-toolkit            3.0.48             pyha770c72_1    conda-forge
prompt_toolkit            3.0.48               hd8ed1ab_1    conda-forge
psutil                    6.1.0           py313h63a2874_0    conda-forge
pthread-stubs             0.4               hd74edd7_1002    conda-forge
ptyprocess                0.7.0              pyhd8ed1ab_1    conda-forge
pure_eval                 0.2.3              pyhd8ed1ab_1    conda-forge
pyarrow                   18.1.0          py313h39782a4_0    conda-forge
pyarrow-core              18.1.0          py313hf9431ad_0_cpu    conda-forge
pycparser                 2.22               pyh29332c3_1    conda-forge
pydantic                  2.10.3             pyh3cfb1c2_0    conda-forge
pydantic-core             2.27.1          py313hdde674f_0    conda-forge
pygments                  2.18.0             pyhd8ed1ab_1    conda-forge
pyobjc-core               10.3.2          py313hb6afeec_0    conda-forge
pyobjc-framework-cocoa    10.3.2          py313hb6afeec_0    conda-forge
pyparsing                 3.2.0              pyhd8ed1ab_2    conda-forge
pysocks                   1.7.1              pyha55dd90_7    conda-forge
python                    3.13.1          h4f43103_102_cp313    conda-forge
python-dateutil           2.9.0.post0        pyhff2d567_1    conda-forge
python-dotenv             1.0.1              pyhd8ed1ab_1    conda-forge
python-fastjsonschema     2.21.1             pyhd8ed1ab_0    conda-forge
python-json-logger        2.0.7              pyhd8ed1ab_0    conda-forge
python-multipart          0.0.19             pyhff2d567_1    conda-forge
python-tzdata             2024.2             pyhd8ed1ab_1    conda-forge
python_abi                3.13                    5_cp313    conda-forge
pytz                      2024.1             pyhd8ed1ab_0    conda-forge
pyyaml                    6.0.2           py313h20a7fcf_1    conda-forge
pyzmq                     26.2.0          py313h0e8b002_3    conda-forge
qhull                     2020.2               h420ef59_5    conda-forge
re2                       2024.07.02           hcd0e937_1    conda-forge
readline                  8.2                  h92ec313_1    conda-forge
referencing               0.35.1             pyhd8ed1ab_1    conda-forge
requests                  2.32.3             pyhd8ed1ab_1    conda-forge
rfc3339-validator         0.1.4              pyhd8ed1ab_1    conda-forge
rfc3986-validator         0.1.1              pyh9f0ad1d_0    conda-forge
rich                      13.9.4             pyhd8ed1ab_1    conda-forge
rich-toolkit              0.11.3             pyh29332c3_0    conda-forge
rpds-py                   0.22.3          py313hdde674f_0    conda-forge
scikit-learn              1.6.0           py313hecba28c_0    conda-forge
scipy                     1.14.1          py313hc010ede_2    conda-forge
send2trash                1.8.3              pyh31c8845_1    conda-forge
setuptools                75.6.0             pyhff2d567_1    conda-forge
shellingham               1.5.4              pyhd8ed1ab_1    conda-forge
six                       1.17.0             pyhd8ed1ab_0    conda-forge
skops                     0.10.0             pyhd8ed1ab_0    conda-forge
skore                     0.5.0              pyhd8ed1ab_0    conda-forge
snappy                    1.2.1                h98b9ce2_1    conda-forge
sniffio                   1.3.1              pyhd8ed1ab_1    conda-forge
soupsieve                 2.5                pyhd8ed1ab_1    conda-forge
stack_data                0.6.3              pyhd8ed1ab_1    conda-forge
starlette                 0.41.3             pyha770c72_1    conda-forge
tabulate                  0.9.0              pyhd8ed1ab_2    conda-forge
tenacity                  9.0.0              pyhd8ed1ab_1    conda-forge
terminado                 0.18.1             pyh31c8845_0    conda-forge
threadpoolctl             3.5.0              pyhc1e730c_0    conda-forge
tinycss2                  1.4.0              pyhd8ed1ab_0    conda-forge
tk                        8.6.13               h5083fa2_1    conda-forge
tomli                     2.2.1              pyhd8ed1ab_1    conda-forge
tornado                   6.4.2           py313h90d716c_0    conda-forge
tqdm                      4.67.1             pyhd8ed1ab_0    conda-forge
traitlets                 5.14.3             pyhd8ed1ab_1    conda-forge
typer                     0.15.1             pyhd8ed1ab_0    conda-forge
typer-slim                0.15.1             pyhd8ed1ab_0    conda-forge
typer-slim-standard       0.15.1               hd8ed1ab_0    conda-forge
types-python-dateutil     2.9.0.20241206     pyhd8ed1ab_0    conda-forge
typing-extensions         4.12.2               hd8ed1ab_1    conda-forge
typing_extensions         4.12.2             pyha770c72_1    conda-forge
typing_utils              0.1.0              pyhd8ed1ab_1    conda-forge
tzdata                    2024b                hc8b5060_0    conda-forge
uri-template              1.3.0              pyhd8ed1ab_1    conda-forge
urllib3                   2.2.3              pyhd8ed1ab_1    conda-forge
uvicorn                   0.32.1             pyh31011fe_1    conda-forge
uvicorn-standard          0.32.1               h31011fe_1    conda-forge
uvloop                    0.21.0          py313h63a2874_1    conda-forge
watchfiles                1.0.3           py313hdde674f_0    conda-forge
wcwidth                   0.2.13             pyhd8ed1ab_1    conda-forge
webcolors                 24.11.1            pyhd8ed1ab_0    conda-forge
webencodings              0.5.1              pyhd8ed1ab_3    conda-forge
websocket-client          1.8.0              pyhd8ed1ab_1    conda-forge
websockets                14.1            py313h90d716c_0    conda-forge
widgetsnbextension        4.0.13             pyhd8ed1ab_1    conda-forge
xorg-libxau               1.0.11               hd74edd7_1    conda-forge
xorg-libxdmcp             1.1.5                hd74edd7_0    conda-forge
yaml                      0.2.5                h3422bc3_2    conda-forge
zeromq                    4.3.5                hc1bb282_7    conda-forge
zipp                      3.21.0             pyhd8ed1ab_1    conda-forge
zstandard                 0.23.0          py313hf2da073_1    conda-forge
zstd                      1.5.6                hb46c0d2_0    conda-forge
```

**Comment by glemaitre** [comment]
OK so `skops` has not been released on conda-forge and `skops==0.10` raise an `ImportError`. I think that the code here:

https://github.com/probabl-ai/skore/blob/a59f0378f4d3b741db1c3b39d8c695528d64ca8e/skore/src/skore/item/__init__.py#L36-L44

should be changed such that we trigger a more informative error with the actual issue.

**Comment by thomass-dev** [comment]
To speed skore import and to limit the number of dependencies, we use lazy-imports in item classes.
This is based on the fact that on missing libraries, an `ImportError` is raised: you can't use `PandasDataFrameItem` without a dataframe, so we've assumed that you have in your environment the `pandas` library.

For the case of `skops ==0.10.0` with `scikit-learn ==1.6.0`, the exception raised is an `ImportError` which is indistinguishable from a real missing library exception.

I will think about a new system.

**Comment by glemaitre** [comment]
> To speed skore import and to limit the number of dependencies, we use lazy-imports in item classes.

Something relevant regarding the lazy import is SPEC001: https://scientific-python.org/specs/spec-0001/. Several libraries are implementing a lazy import mechanism. So maybe it could help here.

**Comment by glemaitre** [comment]
There is a single issue but it was hard to follow at first.

`skore` supports `GridSearchCV`.

But because `skops==0.10` will raise an unrelated `ImportError` (a private class has been moved in scikit-learn 1.6) then because of the `suppress(ImportError)`, we fallback to the `NotImplementError` route.

Ideally, it would be less confusing in this case to raise the original `skops` `ImportError`.

So basically only the second item relevant.

**Comment by MarieSacksick** [comment]
skops 0.13 is out now, and so is sklearn 1.7.  
Skops 0.10 has only been out for 5 months, so we can close this issue.
